### DISCUSSION TITLE
Auto Polygon Suggestions

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#1964] [Ideas]
Hi,

I want add some auto polygon suggestions to make our life easier. 

1 - Add method to return triangles:

```
auto autoPolygon = AutoPolygon::generatePolygon(button->getContentSize());
auto triangles = autoPolygon.getTriangles();
```

Today the have a property "triangles" with a FIXME comment.

2 - We can have a method to create a polygon from triangles to don't need do this all:

```
// create based on image
auto autoPolygon = AutoPolygon::generatePolygon("img1.png");
auto triangles = autoPolygon.triangles;

// convert
std::vector<Vec2> points;
for (int i = 0; i < triangles.indexCount; ++i) {
    auto v3 = triangles.verts[triangles.indices[i]].vertices;
    points.push_back(Vec2(v3.x, v3.y));
}

auto physicsBody = PhysicsBody::createPolygon(points.data(), points.size(), PhysicsMaterial(0.1f, 1.0f, 0.0f));
```

to:

```
auto autoPolygon = AutoPolygon::generatePolygon("img1.png");
auto physicsBody = PhysicsBody::createPolygon(autoPolygon, PhysicsMaterial(0.1f, 1.0f, 0.0f));
```

What do you think?

Thanks.

**Comment by aismann** [comment]
Yes why not.
Please add also an example to cpp_test

**Comment by smilediver** [comment]
Adding `PhysicsBody::createPolygon(autoPolygon)` would make physics system dependent on rendering data structures. Maybe it would be better to extend `AutoPolygon` or `PolygonInfo` to return a `std::vector<Vec2>`.

**Reply by paulocoutinhox** [reply]
Ok. Some code that i do for AutoPolygon class:

```cpp
#include "2d/AutoPolygon.h"
#include "poly2tri/poly2tri.h"
#include "base/Director.h"
#include "base/axstd.h"
#include "renderer/TextureCache.h"
#include "renderer/Texture2D.h"
#include "platform/CCImage.h"
#include "2d/Scale9Sprite.h"
#include "clipper2/clipper.h"
#include <algorithm>
#include <math.h>

// [existing code]

NS_AX_BEGIN

std::vector<Vec2> AutoPolygon::generatePointsForPolygon(const Rect& rect = Rect::ZERO, float epsilon = 2.0f, float threshold = 0.05f, float scaleX = 1.0f, float scaleY = 1.0f, const Vec2& anchorPoint = Vec2::ANCHOR_MIDDLE) {
    auto polygonInfo = generateTriangles(rect, epsilon, threshold);
    auto triangles = polygonInfo.triangles;

    // create vector of points
    std::vector<Vec2> points;
    for (int i = 0; i < triangles.indexCount; ++i) {
        auto v3 = triangles.verts[triangles.indices[i]].vertices;
        points.push_back(Vec2(v3.x * scaleX, v3.y * scaleY));
    }

    // calculate anchor offset
    Vec2 anchorOffset(rect.size.width * anchorPoint.x, rect.size.height * anchorPoint.y);

    // adjust points by subtracting anchor offset
    for (auto &point : points) {
        point -= anchorOffset;
    }

    return points;
}

// [existing code]

NS_AX_END
```

What do you think?

**Reply by smilediver** [reply]
Since you're creating `PolygonInfo` and operating on its data, it would be better to have this method inside `PolygonInfo` class.

But I don't think this method does what you need. This simply returns all the triangle vertices, but `PhysicsBody::createPolygon()` needs a point array that go around the whole polygon, ie you need this:
![image](https://github.com/axmolengine/axmol/assets/1411067/cbf32f46-2782-4e28-91b0-caa6d02443d1)
And also the points must be in clockwise order (P case in image), and the polygon itself must be convex.

And I think `PolygonInfo` can probably have concave shapes, shapes with holes, multiple polygons, etc. So I'm not sure it's a simple problem.

**Reply by paulocoutinhox** [reply]
I changed and create a class helper that inherit from AutoPolygon class. But the most hard part is create polygon info from Texture2D, from SpriteFrame, for example. Ref: https://github.com/axmolengine/axmol/discussions/1974
